test/Test/Database/LSMTree/Internal.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Test.Database.LSMTree.Internal (tests) where
+
+import           Control.Exception
+import           Control.Monad (void)
+import           Database.LSMTree.Internal
+import qualified System.FS.API as FS
+import           Test.Tasty
+import           Test.Tasty.HUnit
+import           Test.Util.FS
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal" [
+      testCase "newSession" newSession
+    , testCase "restoreSession" restoreSession
+    , testCase "twiceOpenSession" twiceOpenSession
+    , testCase "sessionDirLayoutMismatch" sessionDirLayoutMismatch
+    , testCase "sessionDirDoesNotExist" sessionDirDoesNotExist
+    ]
+
+newSession :: Assertion
+newSession = withTempIOHasBlockIO "newSession" $ \hfs hbio ->
+    void $ openSession hfs hbio (FS.mkFsPath [])
+
+restoreSession :: Assertion
+restoreSession = withTempIOHasBlockIO "restoreSession" $ \hfs hbio -> do
+    session1 <- openSession hfs hbio (FS.mkFsPath [])
+    closeSession session1
+    void $ openSession hfs hbio (FS.mkFsPath [])
+
+twiceOpenSession :: Assertion
+twiceOpenSession = withTempIOHasBlockIO "twiceOpenSession" $ \hfs hbio -> do
+    void $ openSession hfs hbio (FS.mkFsPath [])
+    try @LSMTreeError (openSession hfs hbio (FS.mkFsPath [])) >>= \case
+      Left (SessionDirLocked _) -> pure ()
+      _ -> assertFailure "Opening a session twice in the same directory \
+                         \should fail with an SessionDirLocked error."
+
+sessionDirLayoutMismatch :: Assertion
+sessionDirLayoutMismatch = withTempIOHasBlockIO "sessionDirLayoutMismatch" $ \hfs hbio -> do
+    FS.createDirectory hfs (FS.mkFsPath ["unexpected-directory"])
+    try @LSMTreeError (openSession hfs hbio (FS.mkFsPath [])) >>= \case
+      Left _ -> pure ()
```

Comment:
This one could theoretically be either `SessionDirMissingActiveDir` or `SessionDirMissingSnapshotsDir`, right? The inconvenience of potentially having to match on both of these here made me realise that usually the difference doesn't matter to a user, either. So it might be a nicer API to have a single constructor `SessionDirMissingDir` or even `SessionDirMalformed` with further information inside?

Reply:
Yeah, the user probably doesn't really care which exception is thrown in which order. I've changed to a single constructor `SessionDirMalformed` carrying an `FsErrorPath`. The path will give some indication of where the maformation is located. The user might benefit from slightly more detailed information, but I think it's also fine as is